storage/src/journal/contiguous/variable.rs

```diff
@@ -0,0 +1,1477 @@
+                let target_pos = first_section * items_per_section;
+
+                info!("crash repair: rewinding locations from {size} to {target_pos}");
+                locations.rewind(target_pos).await?;
+                locations.sync().await?;
+                return Ok((target_pos, target_pos));
+            }
+
+            // data.blobs is empty. This can happen in two cases:
+            // 1. Rewind crash: we completely pruned the data journal but crashed before rewinding
+            //    the locations journal.
+            // 2. The data journal was never opened.
+            if let Some(oldest) = locations.oldest_retained_pos().await? {
+                if oldest < size {
+                    // Locations has unpruned entries but data is gone - repair by pruning
+                    info!("crash repair: pruning locations to {size} (prune-all crash)");
+                    locations.prune(size).await?;
+                    locations.sync().await?;
+                }
+            }
+
+            return Ok((size, size));
+        }
+
+        // === Handle non-empty data journal case ===
+        let (data_oldest_pos, data_size) = {
+            // Data exists -- count items
+            let first_section = *data.blobs.first_key_value().unwrap().0;
+            let last_section = *data.blobs.last_key_value().unwrap().0;
+
+            let oldest_pos = first_section * items_per_section;
+
+            // Invariant 1 on `Variable` guarantees that all sections except possibly the last
+            // are full. Therefore, the size of the journal is the number of items in the last
+            // section plus the number of items in the other sections.
+            let size = (last_section * items_per_section) + items_in_last_section;
+            (oldest_pos, size)
+        };
+        assert_ne!(
+            data_oldest_pos, data_size,
+            "data journal expected to be non-empty"
+        );
+
+        // Align pruning state. We always prune the data journal before the locations journal,
+        // so we validate that invariant and repair crash faults or detect corruption.
+        match locations.oldest_retained_pos().await? {
+            Some(oldest_retained_pos) if oldest_retained_pos < data_oldest_pos => {
+                // Locations behind on pruning -- prune to catch up
+                info!("crash repair: pruning locations journal to {data_oldest_pos}");
+                locations.prune(data_oldest_pos).await?;
+            }
+            Some(oldest_retained_pos) if oldest_retained_pos > data_oldest_pos => {
+                return Err(Error::Corruption(format!(
+                    "locations oldest pos ({oldest_retained_pos}) > data oldest pos ({data_oldest_pos})"
+                )));
+            }
+            Some(_) => {
+                // Both journals are pruned to the same position.
+            }
+            None if data_oldest_pos > 0 => {
```

Comment:
We can hit this case if we never synced locations (shutdown between data.sync() and locations.sync()). I think we need to tolerate it with add_missing_locations?

Reply:
Added handling for this case and a regression test